service/labelstore/service.go

```diff
@@ -0,0 +1,226 @@
+func (s *service) Data() any {
+	return s
+}
+
+// GetOrAddLink is called by a remote_write endpoint component to add mapping and get back the global id.
+func (s *service) GetOrAddLink(componentID string, localRefID uint64, lbls labels.Labels) uint64 {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	// If the mapping doesn't exist then we need to create it
+	m, found := s.mappings[componentID]
+	if !found {
+		m = &remoteWriteMapping{
+			RemoteWriteID: componentID,
+			localToGlobal: make(map[uint64]uint64),
+			globalToLocal: make(map[uint64]uint64),
+		}
+		s.mappings[componentID] = m
+	}
+
+	labelHash := lbls.Hash()
+	globalID, found := s.labelsHashToGlobal[labelHash]
+	if found {
+		m.localToGlobal[localRefID] = globalID
+		m.globalToLocal[globalID] = localRefID
+		return globalID
+	}
+	// We have a value we have never seen before so increment the globalrefid and assign
+	s.globalRefID++
+	s.labelsHashToGlobal[labelHash] = s.globalRefID
+	m.localToGlobal[localRefID] = s.globalRefID
+	m.globalToLocal[s.globalRefID] = localRefID
+	return s.globalRefID
+}
+
+// GetOrAddGlobalRefID is used to create a global refid for a labelset
+func (s *service) GetOrAddGlobalRefID(l labels.Labels) uint64 {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	// Guard against bad input.
+	if l == nil {
+		return 0
+	}
+
+	labelHash := l.Hash()
+	globalID, found := s.labelsHashToGlobal[labelHash]
+	if found {
+		return globalID
+	}
+	s.globalRefID++
+	s.labelsHashToGlobal[labelHash] = s.globalRefID
+	return s.globalRefID
+}
+
+// GetGlobalRefID returns the global refid for a component local combo, or 0 if not found
+func (s *service) GetGlobalRefID(componentID string, localRefID uint64) uint64 {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	m, found := s.mappings[componentID]
+	if !found {
+		return 0
+	}
+	global := m.localToGlobal[localRefID]
+	return global
+}
+
+// GetLocalRefID returns the local refid for a component global combo, or 0 if not found
+func (s *service) GetLocalRefID(componentID string, globalRefID uint64) uint64 {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	m, found := s.mappings[componentID]
+	if !found {
+		return 0
+	}
+	local := m.globalToLocal[globalRefID]
+	return local
+}
+
+// AddStaleMarker adds a stale marker
+func (s *service) AddStaleMarker(globalRefID uint64, l labels.Labels) {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	s.staleGlobals[globalRefID] = &staleMarker{
+		lastMarkedStale: time.Now(),
+		labelHash:       l.Hash(),
+		globalID:        globalRefID,
+	}
+}
+
+// RemoveStaleMarker removes a stale marker
+func (s *service) RemoveStaleMarker(globalRefID uint64) {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	delete(s.staleGlobals, globalRefID)
+}
+
+// staleDuration determines how often we should wait after a stale value is received to GC that value
+var staleDuration = time.Minute * 10
+
+// CheckStaleMarkers is called to garbage collect and items that have grown stale over stale duration (10m)
+func (s *service) CheckStaleMarkers() {
```

Comment:
`CheckStaleMarkers` doesn't sound like a good name IMO - doesn't reflect that this will delete old markers and garbage-collect effectively. We can call it just `GarbageCollection()` as it may serve multiple purposes.

Reply:
How do you feel about CheckAndRemoveStaleMarkers, I feel GarbageCollection is a bit to general at the moment.